clash-prelude/src/Clash/Explicit/Synchronizer.hs

```diff
@@ -97,27 +97,62 @@ dualFlipFlopSynchronizer clk1 clk2 rst en i =
 -- * Asynchronous FIFO synchronizer
 
 fifoMem
-  :: ( KnownDomain wdom
+  :: forall wdom rdom a addrSize
+   . ( KnownDomain wdom
      , KnownDomain rdom
-     , NFDataX a )
+     , NFDataX a
+     , KnownNat addrSize
+     , 1 <= addrSize )
   => Clock wdom
   -> Clock rdom
   -> Enable wdom
-  -> SNat addrSize
+  -> Enable rdom
   -> Signal wdom Bool
   -> Signal rdom (BitVector addrSize)
-  -> Signal wdom (Maybe (BitVector addrSize, a))
+  -> Signal wdom (BitVector addrSize)
+  -> Signal wdom (Maybe a)
   -> Signal rdom a
-fifoMem wclk rclk en addrSize@SNat full raddr writeM =
-  asyncRam
-    wclk rclk en
-    (pow2SNat addrSize)
-    raddr
-    (mux full (pure Nothing) writeM)
-
-ptrCompareT
-  :: SNat addrSize
-  -> (BitVector (addrSize + 1) -> BitVector (addrSize + 1) -> Bool)
+fifoMem wclk rclk wen ren full raddr waddr wdataM =
+  fst $ trueDualPortBlockRam
+    rclk wclk portA portB
+ where
+   portA :: Signal rdom (RamOp (2 ^ addrSize) a)
+   portA = mux (fromEnable ren)
+               (RamRead . unpack <$> raddr)
+               (pure NoOp)
+   portB :: Signal wdom (RamOp (2 ^ addrSize) a)
+   portB = mux (fromEnable wen .&&. fmap not full .&&. fmap isJust wdataM)
+               (RamWrite <$> fmap unpack waddr <*> fmap fromJustX wdataM)
+               (pure NoOp)
+
+readPtrCompareT
+  :: KnownNat addrSize
+  => ( BitVector (addrSize + 1)
+     , BitVector (addrSize + 1)
+     , Bool )
+  -> ( BitVector (addrSize + 1)
+     , Bool )
+  -> ( ( BitVector (addrSize + 1)
+       , BitVector (addrSize + 1)
+       , Bool )
+     , ( Bool
+       , BitVector addrSize
+       , BitVector (addrSize + 1)
+       )
+     )
+readPtrCompareT (bin, ptr, flag) (s_ptr, inc) =
+  ((bin', ptr', flag'), (flag, addr, ptr))
```

Comment:
Didn't you mention wanting to use one register less in the read-addressing, because a TDP BRAM has registered read inputs whereas `asyncRam` does not? Have you tested you still get the next element when `empty` goes from `True` to `False`?

Reply:
Those primes are too easy to miss, IMHO. I used to do `f`, `f'`, `f''`, `f'3`, `f'4`. But now I do `f`, `f0`, `f1`, ... At least they don't look like a speck on your laptop screen. Note: I have many specks on my laptop screen ;-P.

Reply:
Yeah, I do that as well these days, but I didn't want to change the overall style of the module.